cmd/k3s/main.go

```diff
@@ -115,13 +119,27 @@ func extract(dataDir string) (string, error) {
 	// first look for global asset folder so we don't create a HOME version if not needed
 	_, dir := getAssetAndDir(datadir.DefaultDataDir)
 	if _, err := os.Stat(dir); err == nil {
-		logrus.Debugf("Asset dir %s", dir)
 		return dir, nil
 	}
 
 	asset, dir := getAssetAndDir(dataDir)
+	// check if target directory already exists
+	if _, err := os.Stat(dir); err == nil {
+		return dir, nil
+	}
+
+	// acquire a data directory lock
+	os.MkdirAll(filepath.Join(dataDir, "data"), 0755)
+	lockFile := filepath.Join(dataDir, "data", ".lock")
+	logrus.Infof("Acquiring lock file %s", lockFile)
+	lock, err := flock.Acquire(lockFile)
```

Comment:
Would there ever be a case where we'd want to check for this file first? ie a failed or cancelled run?

Reply:
The file will remain on disk after Releasing the lock (is only closed) in case there is another process trying to Acquire.
A left-over lock file will not block the next Acquire call.
Seems to be part of the magic of unix.Flock (or the flock system call in general) that works at a kernel lock level and accepts a file which already exists but has no existing open handles.

Reply:
Awesome. Thanks!